### PR TITLE
Add Self-supervised methods.

### PR DESCRIPTION
**Pull request by MylesBartlett**

### CODE REVIEW

File: bolts/models/self_supervised/moco/lightning.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+    @implements(pl.LightningModule)
+    def configure_optimizers(self) -> optim.Optimizer:
+        optimizer = optim.SGD(
+            self.student.parameters(),
+            self.lr,
+            momentum=self.momentum_sgd,
+            weight_decay=self.weight_decay,
+        )
+        return optimizer
```
Comment (olliethomas) [inline bolts/models/self_supervised/moco/lightning.py:177]:
```suggestion
        optimizer = optim.SGD(
            self.student.parameters(),
            self.lr,
            momentum=self.momentum_sgd,
            weight_decay=self.weight_decay,
        )
        return optimizer
```
```suggestion
        return optim.SGD(
            self.student.parameters(),
            self.lr,
            momentum=self.momentum_sgd,
            weight_decay=self.weight_decay,
        )
```

File: bolts/models/self_supervised/dino/eval.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        target_dim: int,
+        weight_decay: float,
+        lr: float,
+        epochs: int,
+        num_eval_blocks: int = 1,
+    ) -> None:
+        self.epochs = epochs
+        self.num_eval_blocks = num_eval_blocks
+        if isinstance(encoder, VisionTransformer):
+            embed_dim = embed_dim * num_eval_blocks
```
Comment (olliethomas) [inline bolts/models/self_supervised/dino/eval.py:40]:
```suggestion
            embed_dim *= num_eval_blocks
```
```suggestion
            embed_dim = embed_dim * num_eval_blocks
```

File: bolts/models/utils.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+@torch.no_grad()
+def accuracy(logits: Tensor, targets: Tensor) -> Tensor:
+    logits = torch.atleast_1d(logits.squeeze())
+    targets = torch.atleast_1d(targets.squeeze()).long()
+    if len(logits) != len(targets):
+        raise ValueError("'logits' and 'targets' must match in size at dimension 0.")
+    if logits.ndim == 1:
+        preds = (logits > 0).long()
+    else:
+        preds = logits.argmax(dim=1)
```
Comment (olliethomas) [inline bolts/models/utils.py:50]:
```suggestion
    preds = (logits > 0).long() if logits.ndim == 1 else logits.argmax(dim=1)
```

File: bolts/fair/models/erm.py [hunk 4]
```diff
@@ -1,129 +1,83 @@
...
-            s=pd.DataFrame(all_s.detach().cpu().numpy(), columns=["s"]),
-            y=pd.DataFrame(all_y.detach().cpu().numpy(), columns=["y"]),
+            x=pd.DataFrame(torch.rand_like(subgroups_all).detach().cpu().numpy(), columns=["x0"]),
+            s=pd.DataFrame(subgroups_all.detach().cpu().numpy(), columns=["s"]),
+            y=pd.DataFrame(targets_all.detach().cpu().numpy(), columns=["y"]),
         )
 
-        results = em.run_metrics(
-            predictions=em.Prediction(hard=pd.Series(all_preds.argmax(-1).detach().cpu().numpy())),
+        results_dict = em.run_metrics(
```
Comment (olliethomas) [inline bolts/fair/models/erm.py:76]: return this rather than assigning to a variable

File: bolts/models/base.py [hunk 5]
```diff
@@ -56,36 +62,101 @@ def configure_optimizers(
...
+        self._trainer = trainer
+
+    def build(self, datamodule: PBDataModule, *, trainer: pl.Trainer, copy: bool = True) -> None:
+        if copy:
+            datamodule = gcopy(datamodule, deep=False)
+            trainer = gcopy(trainer, deep=True)
+        self._datamodule = datamodule
+        self._trainer = trainer
+        self._build()
+        # Retrieve all child models (attributes inheriting from ModelBase)
```
Comment (MylesBartlett) [inline bolts/models/base.py:99]:
```suggestion
        # Retrieve all child models (attributes inheriting from PBModel)
```

File: bolts/models/base.py [hunk 6]
```diff
@@ -56,36 +62,101 @@ def configure_optimizers(
...
+        # Build all child models
+        for _, child in children:
+            child.build(datamodule=self.datamodule, trainer=self.trainer, copy=False)
+
+    def _build(self) -> None:
+        ...
+
+    @property
+    def num_training_steps(self) -> int:
+        """Total training steps inferred from datamodule and devices."""
```
Comment (MylesBartlett) [inline bolts/models/base.py:113]:
```suggestion
        """Total training steps inferred from the datamodule and trainer."""
```

**Pull request merged**